Desafio 2 Phoenix

AquilesAguiar opened the pull request.

ArthurFreitas committed:
Clarifica processo de entrega de desafios

AquilesAguiar committed:
first commit

ArthurFreitas commented on phoenix/README.md:
````diff
@@ -9,4 +9,48 @@ Ambiente contendo:
+```
+## Running Aplication
+```
+rails server
````
mix phoenix.server *

ArthurFreitas commented on phoenix/README.md:
```diff
@@ -9,4 +9,48 @@ Ambiente contendo:
+
+##GET
+
+`INDEX GET register/`
```
As rotas precisam ser iguais as do primeiro desafio: /registers

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/management/register.ex:
```diff
@@ -0,0 +1,25 @@
+    field :name, :string
+    field :description, :string
+    field :price, :float
+    field :qtd, :float
```
:integer

ArthurFreitas commented on phoenix/app/cad_products_phoenix/test/cad_products_phoenix/management_test.exs:
```diff
@@ -0,0 +1,72 @@
+  describe "register" do
+    alias CadProductsPhoenix.Management.Register
+
+    @valid_attrs %{description: "some description", name: "some name", price: 120.5, qtd: 120.5, sku: "some sku"}
```
qtd 120.5??

AquilesAguiar committed:
Update README.md

AquilesAguiar committed:
Refactoring 1

Corrections in the routes, corrections in the readme and changing the type of the amount of float to integer

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,43 @@
+    with {:ok, %Register{} = register} <- Management.create_register(register_params) do
+      conn
+      |> put_status(:created)
+      |> put_resp_header("location", Routes.register_path(conn, :show, register))
```
Não entendi a necessidade desse redirect (o header location é utilizado em redirects)

AquilesAguiar committed:
Update README.md

fixes in the readme routes

NoahCamargo commented on phoenix/app/cad_products_phoenix/.gitignore:
```diff
@@ -0,0 +1,28 @@
+# Since we are building assets from assets/,
+# we ignore priv/static. You may want to comment
+# this depending on your deployment strategy.
+/priv/static/
```
Adicione o arquivo de tmp ao ignore

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,43 @@
+    render(conn, "index.json", register: register)
+  end
+
+  def create(conn, %{"register" => register_params} ) do
```
Na verdade é um produto que estamos criando, deveria ser `product`

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/views/register_view.ex:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  def render("show.json", %{register: register}) do
+    %{data: render_one(register, RegisterView, "register.json")}
```
Se entra um `register`, é bom sair a mesma coisa

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/management/register.ex:
```diff
@@ -0,0 +1,25 @@
+    register
+    |> cast(attrs, [:sku, :name, :price, :qtd, :description])
+    |> validate_required([:sku, :name, :price, :qtd, :description])
+  end
```
É bom adicionar uma validação de unicidade no sku, já que ele esta sendo a chave primaria. Ou começar a retornar o ID e aceita-lo nos endpoints

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  def create(conn, %{"register" => register_params} ) do
+    with {:ok, %Register{} = register} <- Management.create_register(register_params) do
```
`with` deve ser usado quando se tem mais de um generator: https://elixirschool.com/pt/lessons/basics/control_structures#with-3 . Nesse caso é melhor substituir por outra estrutura de controle.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/config/dev.exs:
```diff
@@ -0,0 +1,53 @@
+# watchers to your application. For example, we use it
+# with webpack to recompile .js and .css sources.
+config :cad_products_phoenix, CadProductsPhoenixWeb.Endpoint,
+  http: [port: 3000],
```
O aplicativo ta servindo na porta 3000, mas  o container do phoenix ta expondo a porta 4000 por isso não dá para acessar tua aplicação fora do container. A solução é expor a porta 3000 no dockerfile e compose do phoenix, ou servir a aplicação na 4000.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,43 @@
+    end
+  end
+
+  def show(conn, %{"id" => id}) do
```
Falta tratar o que acontece quando o produto não é encontrado

ArthurFreitas replied:
no show, update e delete

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/router.ex:
```diff
@@ -0,0 +1,25 @@
+    resources "/registers", RegisterController, only: [:index, :show, :create, :update, :delete]
+  end
+
+  # scope "/api", CadProductsPhoenixWeb do
```
Sempre remova código comentado

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/views/register_view.ex:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  def render("register.json", %{register: register}) do
+    %{sku: register.sku,
```
roda um mix format aqui

ArthurFreitas reviewed (changes requested):
Tem muitos arquivos e dependências que não são necessárias para o projeto: várias configs de postgres, assets de css, migrations, etc. Seria uma boa tu remover e deixar só o que precisa.

AquilesAguiar committed:
Refactoring 2

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/management.ex:
```diff
@@ -0,0 +1,106 @@
+  import Ecto.Query, warn: false
+  alias CadProductsPhoenix.Repo
+
+  alias CadProductsPhoenix.Management.Register
```
Pode deixar os alias juntos e o import separado.

```
  import Ecto.Query, warn: false

  alias CadProductsPhoenix.Repo
  alias CadProductsPhoenix.Management.Register
```

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,57 @@
+    else
+
+      json(conn, %{error: "Product with #{id} not found"})
+
```
Remover esses espaços extras

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  def update(conn, %{"id" => id, "product" => register_params}) do
+    register = Management.get_register!(id)
```
Utilize `plug` para não precisar ficar repetindo essa linha e a do notfound. funciona como o `before_action` do rails

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/router.ex:
```diff
@@ -0,0 +1,13 @@
+    pipe_through :api
+    resources "/registers", RegisterController, only: [:index, :show, :create, :update, :delete]
+  end
+
```
Remover espaço

NoahCamargo commented on phoenix/app/cad_products_phoenix/.gitignore:
```diff
@@ -0,0 +1,30 @@
+# this depending on your deployment strategy.
+/priv/static/
+
+*.tmp
```
Os arquivos de cache ainda estão presentes no PR

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix/management.ex:
```diff
@@ -35,7 +35,9 @@ defmodule CadProductsPhoenix.Management do
 
   """
-  def get_register!(sku), do: Repo.get_by!(Register, sku: sku)
+  def get_register!(id) do
```
A convenção para usar ! no final de uma função é ela levantar uma exceção. Nesse caso o Repo.get não levanta e o get_register não precisa do !.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -25,19 +23,35 @@ defmodule CadProductsPhoenixWeb.RegisterController do
 
-    with {:ok, %Register{} = register} <- Management.update_register(register, register_params) do
-      render(conn, "show.json", register: register)
+    if register do
```
Pode trocar esse if com um case dentro, para um with só e facilitar a leitura.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -11,12 +11,10 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+  def create(conn, %{"product" => register_params} ) do
+    case Management.create_register(register_params) do
+      {:ok, %Register{} = register} -> render(conn, "show.json", register: register)
+      _ -> render(conn, "show.json", %{error: "Failed to create a product"})
```
Esse código falha porque o parametro que tu passou de `%{error: "msg"}` não casa com nenhum padrão em cad_products_phoenix_web/views/register_view.ex. Dá uma lida melhor em https://elixirschool.com/pt/lessons/basics/pattern_matching.

Exemplo de payload que falha:
```
{
    "product":
    {
        "sku": "sku3",
        "qtd": "",
        "description": "Descrição",
        "name": "monica2",
        "price": 10.4
    }
}
```

ArthurFreitas replied:
Mesma coisa ocorre na linha 33 e 49. Esses render iram chamar RegisterView.render/2 com os parâmetros passados e irão falhar.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -25,19 +23,35 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+
+      case Management.update_register(register, register_params) do
+        {:ok, %Register{} = register} -> render(conn, "show.json", register: register)
+        _ -> render(conn, "show.json", %{error: "Failed to update a product #{id}"})
```
Management.update_register/2 e Management.create_register/2 chamam Repo.update e Repo.create respectivamente e ambos retornam `{:error, changeset}` caso algum erro ocorra (https://hexdocs.pm/ecto/Ecto.Repo-callback-update.html#c:update/2-example). O ideal é retornar os erros para quem fez a requisição conserta-la. 

Dica: o modulo ChangesetView já tem um parser de um changeset implementado.

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -25,19 +23,35 @@ defmodule CadProductsPhoenixWeb.RegisterController do
 
   def delete(conn, %{"id" => id}) do
     register = Management.get_register!(id)
+    if register do
```
Substituir por with

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -25,19 +23,35 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+         end
+    else
+
+      json(conn, %{error: "Product with #{id} not found"})
```
Faltou um HTTP code que represente o erro nessa resposta

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -25,19 +23,35 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+    if register do
+      case Management.delete_register(register) do
+        {:ok, %Register{}} -> send_resp(conn, :no_content, "")
+        _ -> render(conn, "show.json", %{error: "Failed to delete a product #{id}"})
```
Match irá falhar como expliquei lá em cima.

AquilesAguiar committed:
Refactoring 3

I implemented the plugs for reusing the search by id, removed unnecessary spaces and redid the controller logic

AquilesAguiar committed:
Update port

Switch port to 4000

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,56 @@
+      send_resp(conn, :no_content, "")
+    end
+
+  end
```
Espaço extra, remover

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,56 @@
+      {:error, changeset} -> render(conn, "error.json", changeset: changeset)
+    end
+
+  end
```
Remover espaço extra

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,56 @@
+
+    case Management.update_register(register, register_params) do
+      {:ok, %Register{} = register} -> render(conn, "show.json", register: register)
+      {:error, changeset} -> render(conn, "error.json", changeset: changeset)
```
No caso "error.json" não existe no modulo de fallback que esta sendo chamado aqui, porém lá em `FallbackController` ele já possui um matchig para erros de changeset

AquilesAguiar committed:
Refactoring 4

Removal of spaces and correction in the error return in create and update

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,54 @@
+
+    case Management.update_register(register, register_params) do
+      {:ok, %Register{} = register} -> render(conn, "show.json", register: register)
+      {:error, %Ecto.Changeset{} = changeset} -> render(conn, "error.json", changeset: changeset)
```
Aqui no caso ira dar erro, pois o "error.json", esta em outro fallback, porém no fallback que esta sendo passado aqui já recebe em um matching o changeset

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  # Plug for get products
+  defp search_product(conn, __s) do
```
Remover um dos underlines `_s`

AquilesAguiar committed:
Refactoring 5

Using fallback to return error and preview json more correctly

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/views/changeset_view.ex:
```diff
@@ -1,6 +1,6 @@
 defmodule CadProductsPhoenixWeb.ChangesetView do
   use CadProductsPhoenixWeb, :view
-
+  alias CadProductsPhoenixWeb.ChangesetView
```
Não precisa desse alias se já está no próprio módulo, além disso ele não está sendo utilizado

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/views/register_view.ex:
```diff
@@ -1,5 +1,7 @@
   use CadProductsPhoenixWeb, :view
+
+  alias CadProductsPhoenixWeb.ChangesetView
   alias CadProductsPhoenixWeb.RegisterView
```
Uma forma de usar o alias é a seguinte:

```
alias CadProductsPhoenixWeb.{ChangesetView, RegisterView}
```

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -6,52 +6,41 @@ defmodule CadProductsPhoenixWeb.RegisterController do
 
   action_fallback CadProductsPhoenixWeb.FallbackController
 
+  plug :search_product when action in [:show, :update, :delete]
```
Essa forma de usar o plug é totalmente valida, mas quando for usar um plug que tiver uso mais geral, voce precisaria criar um modulo de plug, que define o metodo `call` e `init`

AquilesAguiar committed:
remove tmp

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -6,52 +6,41 @@ defmodule CadProductsPhoenixWeb.RegisterController do
+  def delete(conn, _) do
+    register = conn.assigns[:register]
 
+    with {:ok, %Register{} = register} <- Management.delete_register(register) do
```
a variável register não é usada para nada aqui, substituir por:
`with {:ok, %Register{}}`

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -6,52 +6,41 @@ defmodule CadProductsPhoenixWeb.RegisterController do
     render(conn, "index.json", register: register)
   end
 
   def create(conn, %{"product" => register_params} ) do
```
Colocar um underscore na frente do conn, para sinalizar que não é usado dentro da função

ArthurFreitas commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/fallback_controller.ex:
```diff
@@ -6,6 +6,13 @@ defmodule CadProductsPhoenixWeb.FallbackController do
   """
   use CadProductsPhoenixWeb, :controller
 
+  alias CadProductsPhoenix.Management
```
alias não utilizado, remover

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,46 @@
+    if register do
+      assign(conn, :register, register)
+    else
+      send_resp(conn, :not_found, "Product with #{id} not found")
```
Nesse caso, olhando pelo servidor, ele esta tentando executar duas resposta. Para parar a execução aqui, é necessário adicionar o `halt()`

AquilesAguiar replied:
register do
      conn
      |> assign(:register, register)
      |> halt
    else
      conn
      |> put_status(:not_found)
      |> json(%{error: "Product with #{id} not found"})
      |> halt

AquilesAguiar replied:
Ficaria desse jeito ?

AquilesAguiar committed:
Fix some functions

Adding halt in search_product plug, removing unused aliases and variables, changing error message to json and removing chageset view error render

NoahCamargo commented on phoenix/app/cad_products_phoenix/lib/cad_products_phoenix_web/controllers/register_controller.ex:
```diff
@@ -0,0 +1,50 @@
+    register = Management.get_register(id)
+    if register do
+      conn
+      |> assign(:register, register)
```
Quando o pipeline for de apenas uma linha, usa o formato normal da função `assign(conn, :register, register)`

AquilesAguiar committed:
fix assign and conn

AquilesAguiar committed:
remove space
